src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java

```diff
@@ -323,6 +402,9 @@ public Iterator<Path> iterator() {
      * @param listObjectsV2Publisher the publisher that returns objects and common prefixes that are iterated on.
      * @return an iterator for {@code Path}s constructed from the {@code ListObjectsV2Publisher}s responses.
      */
+    //
+    // TODO: shall this be private?
```

Comment:
An alternative way to make this testable would be to refactor it out of this class and into a helper/ utility class as a public method. That will avoid this class being polluted with protected method that isn't central to the function of the class.

Reply:
I actually found the proper (I think) solution: cover the cases by using newDirectoryStream and getting the iterator from it:

```
Iterator<Path> pathIterator =
            provider.newDirectoryStream(Paths.get(URI.create(pathUri+"/")), filter).iterator();
```

HOWEVER, I was not able to make the test on pagination to work. I think I have done it quite right, but I miss how pagination is controlled. Can you please have a look at it (method pathIteratorForPublisher_withPagination). For now I disabled it and added a TODO.